pkg/submitter/submitter.go

```diff
@@ -0,0 +1,580 @@
+	filterLog, err := logging.ApplyFilter(*cfg, ComponentName, logger)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "apply filter logger")
+	}
+	ctx, close := context.WithCancel(ctx)
+	submitter := &Submitter{
+		ctx:              ctx,
+		close:            close,
+		client:           client,
+		proxy:            proxy,
+		cfg:              cfg,
+		resultCh:         make(chan *mining.Result),
+		account:          account,
+		logger:           log.With(filterLog, "component", ComponentName, "pubKey", account.Address.String()[:6]),
+		contractInstance: contractInstance,
+		submitCount: promauto.NewCounter(prometheus.CounterOpts{
+			Namespace:   "telliot",
+			Subsystem:   "mining",
+			Name:        "submit_total",
+			Help:        "The total number of submitted solutions",
+			ConstLabels: prometheus.Labels{"account": account.Address.String()},
+		}),
+		submitFailCount: promauto.NewCounter(prometheus.CounterOpts{
+			Namespace:   "telliot",
+			Subsystem:   "mining",
+			Name:        "submit_fails_total",
+			Help:        "The total number of failed submission",
+			ConstLabels: prometheus.Labels{"account": account.Address.String()},
+		}),
+		submitProfit: promauto.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace:   "telliot",
+			Subsystem:   "mining",
+			Name:        "submit_profit",
+			Help:        "The current submit profit in percents",
+			ConstLabels: prometheus.Labels{"account": account.Address.String()},
+		},
+			[]string{"slot"},
+		),
+		submitCost: promauto.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace:   "telliot",
+			Subsystem:   "mining",
+			Name:        "submit_cost",
+			Help:        "The current submit cost in 1e18 eth",
+			ConstLabels: prometheus.Labels{"account": account.Address.String()},
+		},
+			[]string{"slot"},
+		),
+		submitReward: promauto.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace:   "telliot",
+			Subsystem:   "mining",
+			Name:        "submit_reward",
+			Help:        "The current reward in 1e18 eth",
+			ConstLabels: prometheus.Labels{"account": account.Address.String()},
+		},
+			[]string{"slot"},
+		),
+	}
+
+	return submitter, submitter.resultCh, nil
+}
+
+func (s *Submitter) Start() error {
+	for {
+		select {
+		case <-s.ctx.Done():
+			if s.resultCh != nil {
+				close(s.resultCh)
+			}
+			if s.lastSubmitCncl != nil {
+				s.lastSubmitCncl()
+			}
+			level.Info(s.logger).Log("msg", "submitter shutdown complete")
+			return s.ctx.Err()
+		case result := <-s.resultCh:
+			if s.lastSubmitCncl != nil {
+				s.lastSubmitCncl()
+			}
+			var ctx context.Context
+			ctx, s.lastSubmitCncl = context.WithCancel(s.ctx)
+
+			level.Info(s.logger).Log("msg", "received a solution",
+				"challenge", fmt.Sprintf("%x", result.Work.Challenge),
+				"solution", result.Nonce,
+				"difficulty", result.Work.Challenge.Difficulty,
+				"requestIDs", fmt.Sprintf("%+v", result.Work.Challenge.RequestIDs),
+			)
+			s.blockUntilTimeToSubmit(ctx)
```

Comment:
no longer needed as we do this inside handleSubmit
```suggestion
```

Reply:
Done